lib/SIL/IR/SILFunctionType.cpp

```diff
@@ -235,8 +235,23 @@ IndexSubset *SILFunctionType::getDifferentiabilityResultIndices() {
       resultIndices.push_back(resultAndIndex.index());
   // Check `inout` parameters.
   for (auto inoutParamAndIndex : enumerate(getIndirectMutatingParameters()))
-    if (inoutParamAndIndex.value().getDifferentiability() !=
-        SILParameterDifferentiability::NotDifferentiable)
+    // FIXME(TF-1305): The `getResults().empty()` condition is a hack.
+    //
+    // Currently, an `inout` parameter can either be:
+    // 1. Both a differentiability parameter and a differentiability result.
+    // 2. `@noDerivative`: neither a differentiability parameter nor a
+    //    differentiability result.
+    // However, there is no way to represent an `inout` parameter that:
+    // 3. Is a differentiability parameter but not a differentiability result.
+    // 4. Is a differentiability result but not a differentiability parameter.
+    //
+    // See TF-1305 for solution ideas. For now, `@noDerivative` `inout`
+    // parameters are not treated as differentiability results, unless the
+    // original function has no formal results, which case all `inout`
+    // parameters are treated as differentiability results.
```

Comment:
Quick paste of solution ideas from TF-1305:

* Add new SILParameterDifferentiability kinds, e.g. `InoutParameterNotDifferentiableResult` (case 3) and `InoutParameterNotDifferentiableParameter` (case 4).
  * Consider whether this needs to be exposed in AST `@differentiable` function types.
* Find some way to store parameter/result differentiability in `SILFunctionType` instead of individual `SILParameterInfo`/`SILResultInfo`.